lib/src/util/condition_scope_visitor.dart

```diff
@@ -0,0 +1,307 @@
+      environment.add(e);
+    }
+  }
+
+  void addAll(Iterable<_ExpressionBox> expressions) {
+    environment.addAll(expressions);
+  }
+
+  Iterable<Expression> getExpressions(Iterable<Element> elements, bool value) {
+    final expressions = <Expression>[];
+    _recursiveGetExpressions(this, expressions, elements, value);
+    return expressions;
+  }
+
+  Iterable<_ExpressionBox> getUndefinedExpressions() =>
+      environment.where((e) => e is _UndefinedExpression);
+
+  static void _recursiveGetExpressions(ConditionScope conditionScope,
+      List<Expression> expressions, Iterable<Element> elements, bool value) {
+    if (conditionScope == null) {
+      return;
+    }
+    for (final element in conditionScope.environment.reversed) {
+      if (element.haveToStop(elements)) {
+        return;
+      }
+      if (element is _ConditionExpression && element.value == value) {
+        expressions.add(element.expression);
+      }
+    }
+    _recursiveGetExpressions(
+        conditionScope.outer, expressions, elements, value);
+  }
+}
+
+abstract class ConditionScopeVisitor extends UnifyingAstVisitor {
+  ConditionScope outerScope;
+
+  Iterable<Expression> getFalseExpressions(Iterable<Element> elements) =>
+      _getExpressions(elements, value: false);
+
+  Iterable<Expression> getTrueExpressions(Iterable<Element> elements) =>
+      _getExpressions(elements);
+
+  @override
+  visitAssignmentExpression(AssignmentExpression node) {
+    _addElementToEnvironment(new _UndefinedExpression(_getLeftElement(node)));
+    node.visitChildren(this);
+  }
+
+  @override
+  visitBlockFunctionBody(BlockFunctionBody node) {
+    _addScope();
+    _addElementToEnvironment(new _UndefinedAllExpression());
+    node.visitChildren(this);
+    _removeLastScope();
+  }
+
+  void visitCondition(Expression node);
+
+  @override
+  visitDoStatement(DoStatement node) {
+    _addScope();
+    visitCondition(node.condition);
+    node.visitChildren(this);
+    _propagateUndefinedExpressions(_removeLastScope());
+  }
+
+  @override
+  visitEmptyFunctionBody(EmptyFunctionBody node) {
+    _addScope();
+    _addElementToEnvironment(new _UndefinedAllExpression());
+    node.visitChildren(this);
+    _removeLastScope();
+  }
+
+  @override
+  visitExpressionFunctionBody(ExpressionFunctionBody node) {
+    _addScope();
+    _addElementToEnvironment(new _UndefinedAllExpression());
+    node.visitChildren(this);
+    _removeLastScope();
+  }
+
+  @override
+  visitForStatement(ForStatement node) {
+    _addScope();
+    _addTrueCondition(node.condition);
+    node.variables?.accept(this);
+    node.initialization?.accept(this);
+    visitCondition(node.condition);
+    node.condition?.accept(this);
+    _addTrueCondition(node.condition);
+    node.updaters.accept(this);
+    node.body?.accept(this);
+    _propagateUndefinedExpressions(_removeLastScope());
+    bool hasBreak = DartTypeUtilities
+        .traverseNodesInDFS(node, excludeCriteria: _isLoopStatement)
```

Comment:
About this:

```dart
while (...) {
  void func() {
    while (...) {
      break;
    }
  }
}
```

The first while search for breaks excluding other for and while, so in that case, the break won't be see because is inside other while, an option would be something.

```dart
while (...) {
  void func() {
    break;
  }
}
```

But that is not possible, the analyzer would say 'A break statement cannot be used outside of a loop or switch statement'